whois/src/main.rs

```diff
@@ -15,173 +15,435 @@
+use serde::{Deserialize, Serialize};
+use serde_binary::binary_stream::Endian;
+use std::{
+	collections::{HashMap, HashSet},
+	fs::{self, File},
+	io::Write,
+};
+use subp2p_explorer::util::{crypto::sr25519, p2p::get_peer_id};
+use subp2p_explorer_cli::commands::authorities::PeerDetails;
 
 #[derive(Clone, Debug, Parser)]
 #[clap(author, version, about = "Simple telemetry feed")]
-struct TelemetryOptions {
+struct WhoIsOptions {
 	#[clap(subcommand)]
 	command: WhoisCommand,
 	/// Web-Socket URLs of a relay chain node.
 	#[clap(long)]
 	pub ws: String,
-	/// Web-Socket URL of a telemetry backend
-	#[clap(long)]
-	pub feed: String,
 	/// Name of a chain to connect
 	#[clap(long)]
 	pub chain: Option<String>,
 	#[clap(flatten)]
 	pub verbose: init::VerbosityOptions,
 	#[clap(flatten)]
 	pub retry: utils::RetryOptions,
+	/// The session index used for computing the validator indicies.
+	#[clap(long)]
+	pub session_index: u32,
+	/// An optional block hash to fetch the queued keys at, otherwise we use the queued keys at the current block.
+	#[clap(long)]
+	pub queued_keys_at_block: Option<H256>,
+	/// Tells if we should update the p2p cache.
+	///
+	/// Note building the p2p cache takes around 10 to 15 minutes
+	#[clap(long)]
+	pub update_p2p_cache: bool,
+	/// Hex-encoded genesis hash of the chain.
+	///
+	/// For example, "781e4046b4e8b5e83d33dde04b32e7cb5d43344b1f19b574f6d31cbbd99fe738"
+	#[clap(long, short)]
+	genesis: String,
+	/// Bootnodes of the chain, must contain a multiaddress together with the peer ID.
+	/// For example, "/ip4/127.0.0.1/tcp/30333/ws/p2p/12D3KooWEyoppNCUx8Yx66oV9fJnriXwCcXwDDUA2kj6vnc6iDEp".
+	#[clap(long, use_value_delimiter = true, value_parser)]
+	bootnodes: Vec<String>,
+	/// The number of seconds the discovery process should run for.
+	#[clap(long, short, value_parser = parse_duration)]
+	timeout: std::time::Duration,
+	/// The address format name of the chain.
+	/// Used to display the SS58 address of the authorities.
+	///
+	/// For example:
+	/// - "polkadot" for Polkadot
+	/// - "substrate" for Substrate
+	/// - "kusama" for Kusama
+	///
+	/// Used to display the SS58 address of the authorities.
+	#[clap(long, short)]
+	address_format: String,
+}
+
+fn parse_duration(arg: &str) -> Result<std::time::Duration, std::num::ParseIntError> {
+	let seconds = arg.parse()?;
+	Ok(std::time::Duration::from_secs(seconds))
 }
 
 #[derive(Clone, Debug, Subcommand)]
 enum WhoisCommand {
-	Account(AccountOptions),
-	Session(SessionOptions),
+	/// Display information about a validator by its SS58-formated account address.
+	ByAccount(AccountOptions),
+	/// Display information about a validator by its index in the para_session.
+	ByValidatorIndex(SessionOptions),
+	/// Display information about a validator by its authority discovery key.
+	ByAuthorityDiscovery(AuthorithyDiscoveryOptions),
+	/// Display information about a validator by its peer id.
+	ByPeerId(PeerIdOptions),
+	/// Display information about all validators.
+	DumpAll,
```

Comment:
Can't it be only one command we need and the cmd-find the needed in the output?

Reply:
I don't have a strong opinion here, could you spell out how the command line calls would look like for each case ? 

For example what I wanted to achieve is something like this  `..... by-validator-index 295 293` where you can copy-paste a string of numbers from other sources and just append them at the end of the command, which I think makes things faster than having to pass them like this: `.... --validator-index 295 --validator-index 293`.